### ISSUE TITLE
Good work!

### ISSUE THREAD

**Issue by huixiancheng** [Guan23/CPGNet#1]
更感兴趣的是P2G 和 G2P的部分，看起来像是cuda写的。不知道作者啥时候能开源

**Comment by Guan23** [comment]
P2G和G2P那两部分其实我也写了，在我自己本地，RV的投影方式很常用了，可以参照RangeNet++啊SalsaNext啊一大堆，就是单独写一个记录投影前点云索引的list列表，我自己写了BEV的投影，那个类框架还是直接扒的RangeNet++，我懒得自己搭类框架了，不过后来我跟作者聊了一下，我对三个相对角的中心理解错了，而且作者又改动了一版，把原始点升到7维而不是9维了，作者说今年下半年会开源，等一等吧。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Guan23/CPGNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年7月8日(星期五) 下午4:15
***@***.***&gt;;
***@***.***&gt;;
主题:&nbsp;[Guan23/CPGNet] Good work! (Issue #1)





 
更感兴趣的是P2G 和 G2P的部分，看起来像是cuda写的。不知道作者啥时候能开源

**Comment by huixiancheng** [comment]
谢谢hhhh. 从共一作者另外的这个[Repo](https://github.com/GangZhang842/SMVF) 里看得更明白了:laughing: .
没想到是F.grid_sample
~~~
class BilinearSample(nn.Module):
    def __init__(self, in_dim, scale_rate):
        super(BilinearSample, self).__init__()
        self.scale_rate = scale_rate
    
    def forward(self, grid_feat, grid_coord):
        '''
        Input:
            grid_feat, (BS, C, H, W)
            grid_coord, (BS, N, 2, S)
        Output:
            pc_feat, (BS, C, N, S)
        '''
        H = grid_feat.shape[2]
        W = grid_feat.shape[3]

        grid_sample_x = (2 * grid_coord[:, :, 1] * self.scale_rate[1] / (W - 1)) - 1 #(BS, N, S)
        grid_sample_y = (2 * grid_coord[:, :, 0] * self.scale_rate[0] / (H - 1)) - 1 #(BS, N, S)

        grid_sample_2 = torch.stack((grid_sample_x, grid_sample_y), dim=-1) #(BS, N, S, 2)
        pc_feat = F.grid_sample(grid_feat, grid_sample_2, mode='bilinear', padding_mode='zeros', align_corners=True) #(BS, C, N, S)
        return pc_feat
~~~

**Comment by huixiancheng** [comment]
不过P2G好像是cuda写的？
https://github.com/GangZhang842/SMVF/blob/22ed741b057f5ed3d465389dd81545b89ca6171c/models/mvf_vfe.py#L112
https://github.com/GangZhang842/SMVF/blob/master/deep_point/setup.py
